libs/fmt/fmt.c

```diff
@@ -418,6 +422,41 @@ DEFINE_PRIM(_I32, ogg_tell, _OGG);
 DEFINE_PRIM(_BOOL, ogg_seek, _OGG _I32);
 DEFINE_PRIM(_I32, ogg_read, _OGG _BYTES _I32 _I32);
 
+/* ----------------------------------------------- SOUND : MP3 ------------------------------------------------ */
+
+typedef struct _fmt_mp3 fmt_mp3;
+struct _fmt_mp3 {
+	char *bytes;
+	int size;
+	mp3dec_t dec;
+	mp3dec_frame_info_t info;
+	mp3d_sample_t pcm[MINIMP3_MAX_SAMPLES_PER_FRAME];
+};
+
+HL_PRIM fmt_mp3 *HL_NAME(mp3_open)( char *bytes, int size ) {
+	fmt_mp3 *o = (fmt_mp3*)hl_gc_alloc_raw(sizeof(fmt_mp3));
+	o->bytes = bytes;
+	o->size = size;
+	mp3dec_init(&o->dec);
+	return o;
+}
+
+HL_PRIM int HL_NAME(mp3_decode_frame)( fmt_mp3 *o, int position, char *output, int size, int offset ) {
+	int ret = -1;
+	hl_blocking(true);
+	ret = mp3dec_decode_frame(&o->dec, o->bytes + position, o->size - position, o->pcm, &o->info);
+	// TODO: Without memcpy?
+	// TODO: Safe write
+	memcpy( (void *)(output + offset), (void *)o->pcm, sizeof(o->pcm) );

```

Comment:
I don't think sizeof(o->pcm) is the best here. You should only copy the amount of frame_bytes as specified in documentation. Please also handle errors (ret = 0) and return -1 without copying if the size to copy exceed `size`

Reply:
`dec.frame_bytes` refer to amount of bytes read from input buffer, e.g. assumed next frame offset. To copy only the amount of decoded samples (`pcm` uses 2-channel maximum size) it would need to calculate them trough `ret * o->info.channels * 4`.
As of errors:
`-1` when `position` is out of bounds of file bounds and when output buffer is too small.
`0` when no mp3 data was decoded.

Reply:
Then output size should be as big as pcm, because it's impossible to know the sample count in frame before decoding it (or doing extra work on finding and reading header and interpreting it to detect amount of samples)